file_writer_control/WorkerCommandChannel.py

```diff
@@ -0,0 +1,95 @@
+        Attempt to start a file-writing job by finding an available (idle) worker and sending it a write command.
+        Will re-try to send a write command until a timeout has been reached. This member function is not blocking.
+        :param job: The job to be started.
+        :return: A CommandHandler for (more) easily checking the outcome of the start command.
+        """
+        thread_kwargs = {"do_job": job, "stop_event": self._stop_event}
+        self.command_channel.add_job_id(job.job_id)
+        self.command_channel.add_command_id(job.job_id, job.job_id)
+        temp_thread = threading.Thread(
+            target=self._start_job_thread_function, daemon=True, kwargs=thread_kwargs
+        )
+        temp_thread.start()
+        self.start_job_threads.append(temp_thread)
+        return CommandHandler(self.command_channel, job.job_id)
+
+    def get_idle_workers(self) -> List[WorkerStatus]:
+        """
+        Get a list of workers that has the (known) state idle. May be an empty list. Idle status may be outdated.
+        :return: List of workers.
+        """
+        list_of_workers = self.list_known_workers()
+        list_of_idle_workers = []
+        for worker in list_of_workers:
+            if worker.state == WorkerState.IDLE:
+                list_of_idle_workers.append(worker)
+        return list_of_idle_workers
+
+    def _start_job_thread_function(self, do_job: WriteJob, stop_event: threading.Event):
+        """
+        Attempt to find an idle worker (file-writer) and give it a file-writing job.
+        :param do_job: The write job to start.
+        :param stop_event: Used for stopping the thread before it has finished by itself.
+        """
+        job_started_time = start_time = time.time()
+
+        waiting_to_send_job = True
+
+        while start_time + START_JOB_TIMEOUT > time.time() and not stop_event.is_set():
+            if waiting_to_send_job:
+                list_of_idle_workers = self.get_idle_workers()
+                if len(list_of_idle_workers) > 0:
+                    used_worker = list_of_idle_workers[
+                        randrange(len(list_of_idle_workers))
+                    ]
+                    do_job.service_id = used_worker.service_id
+                    self.message_producer.send(
+                        self.command_topic, do_job.get_start_message()
+                    )
+                    waiting_to_send_job = False
+                    job_started_time = time.time()
+            else:
+                list_of_jobs = self.command_channel.list_jobs()
+                for job in list_of_jobs:
+                    if job.job_id == do_job.job_id:
+                        if job.state == JobState.WRITING or job.state == JobState.DONE:
+                            return
+                        elif job_started_time + SEND_JOB_TIMEOUT < time.time():
+                            waiting_to_send_job = True
+                        break
+            time.sleep(1)
```

Comment:
Why do we need to sleep?

Reply:
None of the other calls in this function are blocking and we do not want to hammer the CPU.

Reply:
Do we have an observation indeed that is the case ie we hammer the CPU without this?

Reply:
Well, you can take my word for it or test it yourself. :-)

Reply:
It is not about me taking your word :) I am sure that there will be some pressure on the CPU. My problem is the magic numbers like `1 second`, `5 seconds` without any hard evidence that these are going to hold for every case in the future that we may experience. These are not easily debug-able. 

Even if we assume, for some case, we found out these magic numbers should be `0.9 instead of 1` and `4.5 instead of 5`, tuning according to problem at hand is neither efficient nor maintainable. 

A more Pythonic (or engineering for that matter) way, if we insist on using `time.sleep()`, is to create a [context manager](https://docs.python.org/3/library/contextlib.html) like `with time(t):` putting this into a utilities module and wrapping the places in need with this manager. In this case we provide `t` as a parameter with values `1, 5` whatever. One can then create a built-in exception within the manager and thus any related bug can be traced back. Furthermore, we then have a reusable module concerning these matters.

If you believe my suggestion is not suitable, please add Matt as a second reviewer for another opinion. I am currently not inclined to put magic numbers to make Jenkins or a CPU happy :)

Reply:
If you look at what the function does and compare time values of the timeouts I think that you can realise the hammering the CPU (i.e. removing the sleep statement) is a complete waste of resources. It might also cause problems elsewhere in the application due to the Python global interpreter lock.
There is one more piece of information that might be helpful, the status repetition rate in the file-writer is by default set to 1 or two seconds (I do not remember which right now). So, assuming that the default status update rate is used in the file-writer is not changed, having the sleep statement on the same order of magnitude is appropriate. However, the exact value matters little in this particular case.
Of curse, the code could be re-written to not use polling. However, that is also a waste of resources (i.e. my development time) for something that is really not time critical.
Though you have not asked for it, I can put the value in a variable (as I probably should have done from the beginning). But spending time investigating the optimal value of that variable is just a waste of resources.

Reply:
Sure. I agree looking for an exact value is not necessary. What I am suggesting is to create a traceable code for maintainability which is not a waste of resources. A context manager is way to go here. Let us add @mattclarke  as a reviewer and get a second opinion. If he agrees with your points then you and Matt can merge the code.

Reply:
See the basic implementation here https://github.com/ess-dmsc/file-writer-control/pull/15